File: studio/gates.py
"""Approval gate polling for studio runs."""

import json
import os
import time

from harness.packets import atomic_write_json


def _approval_path(run_dir, gate):
    return os.path.join(run_dir, "approvals", f"{gate}.json")


def wait(run_dir, gate, timeout_s, jw):
    """Poll for human approval; never auto-approves."""
    jw.emit("gate_open", gate=gate)
    deadline = time.time() + timeout_s
    path = _approval_path(run_dir, gate)
    while time.time() < deadline:
        if os.path.exists(path):
            with open(path, encoding="utf-8") as f:
                data = json.load(f)
            jw.emit("gate_result", gate=gate, decision=data.get("decision", ""))
            return data
        time.sleep(0.5)
    jw.emit("gate_timeout", gate=gate)
    return {"decision": "timeout"}


def write_approval(run_dir, gate, decision, payload=None):
    """Atomically write an approval file."""
    os.makedirs(os.path.join(run_dir, "approvals"), exist_ok=True)
    atomic_write_json(
        _approval_path(run_dir, gate),
        {
            "decision": decision,
            "payload": payload or {},
            "ts": time.time(),
        },
    )


def is_gate_open(journal_path, gate):
    """True if journal has an unanswered gate_open for this gate."""
    from harness.packets import journal_read

    journal = journal_read(journal_path)
    last_open = last_result = None
    for e in journal:
        if e.get("event") == "gate_open" and e.get("gate") == gate:
            last_open = e
        if e.get("event") == "gate_result" and e.get("gate") == gate:
            last_result = e
    if not last_open:
        return False
    if not last_result:
        return True
    return last_open.get("seq", 0) > last_result.get("seq", 0)

File: src/server/gates.ts
/** Port of backend/studio/gates.py — wait() for driver polling. */
import fs from "node:fs";
import path from "node:path";
import { atomicWriteJson, journalRead, type JournalEvent, type JournalWriter } from "./packets";

function approvalPath(runDir: string, gate: string): string {
  return path.join(runDir, "approvals", `${gate}.json`);
}

export function writeApproval(
  runDir: string,
  gate: string,
  decision: string,
  payload: Record<string, unknown> | null = null,
): void {
  fs.mkdirSync(path.join(runDir, "approvals"), { recursive: true });
  atomicWriteJson(approvalPath(runDir, gate), {
    decision,
    payload: payload ?? {},
    ts: Date.now() / 1000,
  });
}

export function isGateOpen(journalPath: string, gate: string): boolean {
  const journal = journalRead(journalPath);
  let lastOpen: JournalEvent | null = null;
  let lastResult: JournalEvent | null = null;
  for (const e of journal) {
    if (e.event === "gate_open" && e.gate === gate) lastOpen = e;
    if (e.event === "gate_result" && e.gate === gate) lastResult = e;
  }
  if (!lastOpen) return false;
  if (!lastResult) return true;
  return (lastOpen.seq ?? 0) > (lastResult.seq ?? 0);
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted) {
      reject(new Error("aborted"));
      return;
    }
    const timer = setTimeout(resolve, ms);
    signal?.addEventListener("abort", () => {
      clearTimeout(timer);
      reject(new Error("aborted"));
    });
  });
}

export async function waitGate(
  runDir: string,
  gate: string,
  timeoutS: number,
  jw: JournalWriter,
  signal?: AbortSignal,
): Promise<Record<string, unknown>> {
  jw.emit("gate_open", { gate });
  const deadline = Date.now() / 1000 + timeoutS;
  const path = approvalPath(runDir, gate);
  while (Date.now() / 1000 < deadline) {
    signal?.throwIfAborted?.();
    if (fs.existsSync(path)) {
      const data = JSON.parse(fs.readFileSync(path, "utf8")) as Record<string, unknown>;
      jw.emit("gate_result", { gate, decision: data.decision ?? "" });
      return data;
    }
    try {
      await sleep(500, signal);
    } catch {
      jw.emit("gate_timeout", { gate });
      return { decision: "timeout" };
    }
  }
  jw.emit("gate_timeout", { gate });
  return { decision: "timeout" };
}
